src/server/channel_store.h

```diff
@@ -34,39 +46,100 @@ class ChannelStore {
-      return lhs.thread_id < rhs.thread_id;
-    }
+    static bool ByThread(const Subscriber& lhs, const Subscriber& rhs);
+
+    ConnectionContext* conn_cntx;
+    util::fibers_ext::BlockingCounter borrow_token;  // to keep connection alive
+    uint32_t thread_id;
+    std::string pattern;  // non-empty if registered via psubscribe
   };
 
-  void AddSub(std::string_view channel, ConnectionContext* me, uint32_t thread_id);
-  void RemoveSub(std::string_view channel, ConnectionContext* me);
+  // Centralized controller to prevent overlaping updates.
+  struct ControlBlock {
+    void Destroy();
 
-  void AddPatternSub(std::string_view pattern, ConnectionContext* me, uint32_t thread_id);
-  void RemovePatternSub(std::string_view pattern, ConnectionContext* me);
+    ChannelStore* most_recent;
+    ::boost::fibers::mutex update_mu;  // locked during updates.
+  };
+
+  // Initialize ChannelStore bound to specific control block.
+  ChannelStore(ControlBlock* cb);
 
-  std::vector<Subscriber> FetchSubscribers(std::string_view channel);
+  // Fetch all subscribers for channel, including matching patterns.
+  std::vector<Subscriber> FetchSubscribers(std::string_view channel) const;
 
   std::vector<std::string> ListChannels(const std::string_view pattern) const;
   size_t PatternCount() const;
 
  private:
   using ThreadId = unsigned;
+
+  // Subscribers for a single channel/pattern.
   using SubscribeMap = absl::flat_hash_map<ConnectionContext*, ThreadId>;
 
-  struct ChannelMap : absl::flat_hash_map<std::string, std::unique_ptr<SubscribeMap>> {
+  // Wrapper around atomic pointer that allows copying and moving.
+  // Made to overcome restrictions of absl::flat_hash_map.
+  // Copy/Move don't need to be atomic with RCU.
+  struct UpdatablePointer {
+    UpdatablePointer(SubscribeMap* sm) : ptr{sm} {
+    }
+
+    UpdatablePointer(const UpdatablePointer& other);
+
+    SubscribeMap* Get() const;
+    void Set(SubscribeMap* sm);
+
+    SubscribeMap* operator->();
+    const SubscribeMap& operator*() const;
+
+   private:
+    std::atomic<SubscribeMap*> ptr;
+  };
+
+  // Subscriber maps for channels/pointers.
+  struct ChannelMap : absl::flat_hash_map<std::string, UpdatablePointer> {
     void Add(std::string_view key, ConnectionContext* me, uint32_t thread_id);
     void Remove(std::string_view key, ConnectionContext* me);
+
+    // Delete stored all SubscribeMap pointers.
+    void DeleteAll();
   };
 
+ private:
+  ChannelStore(ChannelMap* channels, ChannelMap* patterns, ControlBlock*);
+
+  // Size of underlying SubscribeMapf or key, or 0 if not present.
+  unsigned SlotSize(bool pattern, std::string_view key);
+
   static void Fill(const SubscribeMap& src, const std::string& pattern,
                    std::vector<Subscriber>* out);
 
-  mutable folly::RWSpinLock lock_;
-  ChannelMap channels_;
-  ChannelMap patterns_;
+  ChannelMap* channels_;
+  ChannelMap* patterns_;
+  ControlBlock* control_block_;
+};
```

Comment:
That's a lot of indirections as its easier to work it this way, but to flatten it we can:
1. Store the ChannelStore in threads by value (3x8)
2. Store the ChannelMaps without pointer and move them around with ugly memcpy's
3. Store a single ChannelStore and make the ChannelMap* pointers atomic. We still need to dispatch to the shard_set to allow deletes (as we don't have hazard pointers)